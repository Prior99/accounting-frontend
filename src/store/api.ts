import { store } from ".";

declare var baseUrl: string;

type HTTPMethod = "DELETE" | "GET" | "PUT" | "POST";

export async function api(url: string, body?: any, method: HTTPMethod = "GET"): Promise<any> {
    const headers = new Headers();
    if (store.login.loggedIn) {
        headers.append("authorization", store.login.authToken);
    }
    headers.append("content-type", "application/json");
    const fullUrl = `${window.location.protocol}//${baseUrl}${url}`;
    try {
        const response = await fetch(fullUrl, {
            method,
            headers,
            body: JSON.stringify(body),
        });
        const jsonResponse = (await response.json());
        const okay = response.status >= 200 && response.status <= 299;
        if (!okay) {
            store.error.apiError(`Fetching from url "${fullUrl}" resulted in error: "${jsonResponse.message}"`);
            return { okay };
        }
        return {
            okay,
            data: jsonResponse.data,
        };
    } catch (err) {
        store.error.apiError(`Failed to fetch from url "${fullUrl}".`, err);
        return { okay: false };
    }
}
