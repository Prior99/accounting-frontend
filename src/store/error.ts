import { observable, computed, action } from "mobx";

export interface DisplayableError {
    dismissed: boolean;
    message: string;
    origin: "api" | "internal";
    error?: Error;
}

export class ErrorStore {
    constructor() {
        this.apiError = this.apiError.bind(this);
        this.internalError = this.internalError.bind(this);
        this.dismissCurrent = this.dismissCurrent.bind(this);
    }

    @observable public errors: DisplayableError[] = [];

    @computed public get current() {
        return this.errors.find(error => !error.dismissed);
    }

    @action
    public dismissCurrent() {
        if (typeof this.current === "undefined") {
            return;
        }
        this.current.dismissed = true;
    }

    @action
    public apiError(message: string, error?: Error) {
        this.errors.push({ dismissed: false, message, origin: "api", error });
    }

    @action
    public internalError(message: string, error?: Error) {
        this.errors.push({ dismissed: false, message, origin: "internal", error });
    }
}
