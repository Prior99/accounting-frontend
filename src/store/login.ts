import { observable, computed, action } from "mobx";
import { api } from "./api";
import { bind } from "bind-decorator";

export class LoginStore {
    @observable public authToken: string;
    @observable public failed = false;

    @computed
    public get loggedIn() {
        return typeof this.authToken !== "undefined";
    }

    @bind
    @action
    public async onLogin(email: string, password: string) {
        const result = await api("/auth/login", { email, password }, "POST", true);
        const { data } = result;
        if (result.okay && data.auth_token) {
            this.authToken = data.auth_token;
        } else {
            this.failed = true;
        }
    }
}
