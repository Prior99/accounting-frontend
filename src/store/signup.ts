import { observable, computed, action } from "mobx";
import { bind } from "bind-decorator";
import { ApiStore } from ".";
import { History } from "history";
import { Store } from "./store";

export class SignupStore extends Store {
    @observable
    public signupResult: Boolean;

    @bind @action
    public async doSignup(email: string, password: string) {
        const body = { email, password };
        const response = await this.api.call("doSignup", () => this.api.registration.v1AuthRegisterPost({ body }));
        if (response) {
            await this.api.doLogin(email, password);
            this.signupResult = true;
        } else {
            this.signupResult = false;
        }
    }
}
