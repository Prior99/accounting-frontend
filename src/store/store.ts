import { ApiStore } from "./api";
import { History } from "history";

export class Store {
    protected api: ApiStore;
    protected browserHistory: History;

    constructor(api: ApiStore, browserHistory: History) {
        this.api = api;
        this.browserHistory = browserHistory;
    }
}
