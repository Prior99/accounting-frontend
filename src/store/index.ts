import { LoginStore } from "./login";
import { SignupStore } from "./signup";
import { ErrorStore } from "./error";
export * from "./error";

export interface Store {
    login: LoginStore;
    error: ErrorStore;
    signup: SignupStore;
}

export let store: Store;

export function createStore(): Store {
    return {
        login: new LoginStore(),
        error: new ErrorStore(),
        signup: new SignupStore()
    };
}

export function resetStore() {
    store = createStore();
}

resetStore();

if (typeof window !== "undefined") {
    (window as any).store = store;
    (window as any).createStore = createStore;
    (window as any).resetStore = resetStore;
}
