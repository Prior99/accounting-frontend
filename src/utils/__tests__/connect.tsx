import * as React from "react";
import { connect } from "..";
import { store, Store, resetStore } from "../../store";
import { create } from "react-test-renderer";

function mapStoreToProps(argStore: Store) {
    return {
        username: argStore.login.username,
    };
}

beforeEach(() => {
    resetStore();
    store.login.username = "connected";
});

test("`connect()` connects the `Store` to a normal component", () => {
    class UnconnectedComponent extends React.Component<{ username: string }, undefined> {
        public render() {
            return (
                <div>{this.props.username}</div>
            );
        }
    }

    const ConnectedComponent = connect(UnconnectedComponent, mapStoreToProps);
    expect(create(<ConnectedComponent />)).toMatchSnapshot();
    expect(create(<UnconnectedComponent username="unconnected" />)).toMatchSnapshot();
});

test("`connect()` connects the `Store` to a stateless component", () => {
    function UnconnectedComponent ({ username }: { username: string }) {
        return (
            <div>{username}</div>
        );
    }

    const ConnectedComponent = connect(UnconnectedComponent, mapStoreToProps);
    expect(create(<ConnectedComponent />)).toMatchSnapshot();
    expect(create(<UnconnectedComponent username="unconnected" />)).toMatchSnapshot();
});
