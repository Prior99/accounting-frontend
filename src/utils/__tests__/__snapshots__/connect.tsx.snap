// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`\`connect()\` connects the \`Store\` to a normal component 1`] = `
<div>
  connected
</div>
`;

exports[`\`connect()\` connects the \`Store\` to a normal component 2`] = `
<div>
  unconnected
</div>
`;

exports[`\`connect()\` connects the \`Store\` to a stateless component 1`] = `
<div>
  connected
</div>
`;

exports[`\`connect()\` connects the \`Store\` to a stateless component 2`] = `
<div>
  unconnected
</div>
`;
